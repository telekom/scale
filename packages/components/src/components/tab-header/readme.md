# scale-tab-header



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute  | Description                           | Type      | Default     |
| ---------- | ---------- | ------------------------------------- | --------- | ----------- |
| `disabled` | `disabled` | True for a disabled Tabnavigation     | `boolean` | `false`     |
| `selected` | `selected` |                                       | `boolean` | `undefined` |
| `small`    | `small`    | True for smaller height and font size | `boolean` | `false`     |
| `styles`   | `styles`   | (optional) Injected CSS styles        | `string`  | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
