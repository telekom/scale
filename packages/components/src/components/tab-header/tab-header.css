/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  outline: none;
  --font-family: var(--telekom-typography-font-family-sans, TeleNeoWeb);
  --font-weight: var(--telekom-typography-font-weight-bold);
  --height-small: 2rem;
  --color: var(--telekom-color-text-and-icon-standard);
  --color-hover: var(--telekom-color-text-and-icon-primary-hovered);
  --color-selected: var(--telekom-color-text-and-icon-primary-standard);
  --background-underline-hover: var(--telekom-color-primary-hovered);
  --background-underline-active: var(--telekom-color-primary-pressed);
  --color-active: var(--telekom-color-text-and-icon-primary-standard);
  --box-shadow-focus: 0 0 0 var(--telekom-line-weight-highlight)
    var(--telekom-color-functional-focus);
  --spacing-right-slotted: var(--telekom-spacing-unit-x2);
  --color-disabled: var(--telekom-color-text-and-icon-disabled);
  --radius: var(--telekom-radius-standard);
}

.tab-header {
  color: var(--color);
  cursor: pointer;
  display: inline-flex;
  outline: none;
  position: relative;
  box-sizing: border-box;
  align-items: center;
  padding-left: 0;
  height: var(--tab-height);
  margin-right: var(--tab-spacing);
  padding-right: var(--tab-spacing);
  font-family: var(--font-family);
  font-size: calc(var(--tab-height) / 2);
  font-weight: var(--font-weight);
  border-radius: var(--tab-radius);
}

.tab-header--small {
  --tab-height: var(--height-small);
}

.tab-header--disabled {
  cursor: auto;
  color: var(--color-disabled);
}
.tab-header--disabled:after {
  cursor: auto;
  background-color: var(--telekom-color-ui-outlined-border-disabled) !important;
}

.tab-header--disabled:active {
  color: var(--color-disabled);
}

.tab-header--disabled:hover {
  color: var(--color-disabled) !important;
}

.tab-header--disabled:hover:after {
  background-color: var(--telekom-color-ui-outlined-border-disabled) !important;
}

.tab-header:after {
  left: 0;
  right: 0;
  width: 100%;
  bottom: 0;
  content: '';
  display: block;
  position: absolute;
  height: var(--tab-border-size);
  background-color: var(--tab-border-color);
}
.tab-header:before {
  left: 0;
  right: 0;
  width: 100%;
  bottom: 0;
  height: 0;
  content: '';
  display: block;
  z-index: 1;
  position: absolute;
  border-top: 1px solid transparent;
}
.tab-header:hover {
  color: var(--color-hover);
}
.tab-header:hover:after {
  background-color: var(--background-underline-hover);
}
.tab-header:active {
  color: var(--color-active);
}
.tab-header:active:after {
  background-color: var(--background-underline-active);
}
.tab-header ::slotted(*) {
  margin-right: var(--spacing-right-slotted);
}
.tab-header--has-focus {
  box-shadow: var(--box-shadow-focus);
}
.tab-header--selected {
  color: var(--color-selected);
}
.tab-header--selected:after {
  border-top-left-radius: var(--radius);
  border-top-right-radius: var(--radius);
  height: var(--tab-border-size-selected);
  background-color: var(--color-selected);
}
.tab-header--selected:before {
  border-top: var(--tab-border-size-selected) solid transparent;
}
