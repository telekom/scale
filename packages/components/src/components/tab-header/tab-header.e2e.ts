/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newE2EPage } from '@stencil/core/testing';

describe('scale-tab-header', () => {
  it('renders', async () => {
    const page = await newE2EPage();
    await page.setContent(
      '<scale-tab-header slot="tab">General</scale-tab-header>'
    );
    const element = await page.find('scale-tab-header');
    expect(element).toHaveClass('hydrated');
  });

  it('should set size for icons', async () => {
    const page = await newE2EPage();
    await page.setContent(`
      <scale-tab-header>
        <scale-icon-home-home></scale-icon-home-home> General
      </scale-tab-header>
    `);
    const element = await page.find('scale-icon-home-home');
    expect(element.getAttribute('size')).toBe('16');
  });
});
