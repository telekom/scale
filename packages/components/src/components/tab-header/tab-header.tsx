/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { Component, h, Prop, Host, Watch, State, Element } from '@stencil/core';
import classNames from 'classnames';

let i = 0;

@Component({
  tag: 'scale-tab-header',
  styleUrl: './tab-header.css',
  shadow: true,
})
export class TabHeader {
  generatedId: number = i++;
  container: HTMLElement;

  @Element() el: HTMLElement;

  /** True for smaller height and font size */
  @Prop() small: boolean = false;
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;
  @Prop() selected: boolean;

  @State() hasFocus: boolean = false;

  @Watch('selected')
  selectedChanged(newValue: boolean) {
    if (newValue === true) {
      // Having focus on the host element, and not on inner elements,
      // is required because screen readers.
      this.el.focus();
    }
    this.updateSlottedIcon();
  }

  /**
   * Find slotted icons, and if any, add the `selected` attribute accordingly.
   */
  updateSlottedIcon() {
    const slot = this.container.querySelector('slot') as HTMLSlotElement;
    if (slot === null) {
      return;
    }
    const children = Array.from(slot.assignedNodes())
      .filter((node) => node.nodeType === 1)
      .filter((node) => node.nodeName.toUpperCase().indexOf('ICON') > -1);
    if (children.length === 0) {
      return;
    }
    const action = this.selected ? 'setAttribute' : 'removeAttribute';
    children.forEach((child) => child[action]('selected', ''));
  }

  render() {
    return (
      <Host
        id={`scale-tab-header-${this.generatedId}`}
        role="tab"
        aria-selected={String(this.selected)}
        tabindex={this.selected ? '0' : '-1'}
        onFocus={() => (this.hasFocus = true)}
        onBlur={() => (this.hasFocus = false)}
      >
        {this.styles && <style>{this.styles}</style>}

        <span
          part={this.getBasePartMap()}
          ref={(el) => (this.container = el)}
          class={this.getCssClassMap()}
        >
          <slot />
        </span>
      </Host>
    );
  }

  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }

  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }

  getCssOrBasePartMap(mode: 'basePart' | 'css') {
    const component = 'tab-header';
    const prefix = mode === 'basePart' ? '' : `${component}--`;

    return classNames(
      component,
      this.selected && `${prefix}selected`,
      this.small && `${prefix}small`,
      this.hasFocus && `${prefix}has-focus`
    );
  }
}
