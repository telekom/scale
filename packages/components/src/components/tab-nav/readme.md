# scale-tab-nav



<!-- Auto Generated Below -->


## Properties

| Property | Attribute | Description                                                                           | Type                 | Default     |
| -------- | --------- | ------------------------------------------------------------------------------------- | -------------------- | ----------- |
| `size`   | `size`    | (optional) size                                                                       | `"large" \| "small"` | `'large'`   |
| `small`  | `small`   | <span style="color:red">**[DEPRECATED]**</span> - size should replace small<br/><br/> | `boolean`            | `false`     |
| `styles` | `styles`  | (optional) Injected CSS styles                                                        | `string`             | `undefined` |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
