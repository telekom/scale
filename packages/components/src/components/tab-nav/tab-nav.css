/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

:host {
  --tab-height: var(--telekom-spacing-unit-x10);
  --tab-spacing: var(--telekom-spacing-unit-x6);
  --tab-radius: var(--telekom-radius-small);
  --tab-border-size: var(--telekom-line-weight-highlight);
  --tab-border-size-selected: var(--telekom-line-weight-bold);
  --tab-border-color: var(--telekom-color-ui-additional-subtle);
}
