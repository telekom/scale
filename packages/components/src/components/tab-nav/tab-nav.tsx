/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import {
  Element,
  Component,
  h,
  Prop,
  Host,
  Listen,
  Watch,
} from '@stencil/core';
import classNames from 'classnames';
import statusNote from '../../utils/status-note';

/**
 * @see https://github.com/GoogleChromeLabs/howto-components/blob/master/elements/howto-tabs/howto-tabs.js
 */

const ARROW_LEFT = 'ArrowLeft';
const ARROW_RIGHT = 'ArrowRight';
const HOME = 'Home';
const END = 'End';

@Component({
  tag: 'scale-tab-nav',
  styleUrl: './tab-nav.css',
  shadow: true,
})
export class TabNav {
  @Element() el: HTMLElement;

  /** True for smaller height and font size in tab headers. */
  /** @deprecated - size should replace small */
  @Prop() small: boolean = false;
  /** (optional) size  */
  @Prop() size: 'small' | 'large' = 'large';
  /** (optional) Injected CSS styles */
  @Prop() styles?: string;

  @Watch('small')
  smallChanged() {
    this.propagateSizeToTabs();
  }

  @Listen('click')
  handleClick(event: MouseEvent) {
    const nextTab = event.target as HTMLScaleTabHeaderElement;
    if (nextTab.getAttribute('role') !== 'tab') {
      return;
    }
    this.selectTab(nextTab);
  }

  @Listen('keydown')
  handleKeydown(event: KeyboardEvent) {
    const target = event.target as HTMLScaleTabHeaderElement;
    let nextTab;

    if (target.getAttribute('role') !== 'tab') {
      return;
    }
    // Do not handle modifier shortcuts typically used by assistive technology
    if (event.altKey) {
      return;
    }

    switch (event.key) {
      case ARROW_LEFT:
        nextTab = this.getPreviousTab();
        break;
      case ARROW_RIGHT:
        nextTab = this.getNextTab();
        break;
      case HOME:
        nextTab = this.getFirstTab();
        break;
      case END:
        nextTab = this.getLastTab();
        break;
      default:
        return;
    }

    event.preventDefault();
    this.selectTab(nextTab);
  }

  connectedCallback() {
    if (!this.el.hasAttribute('role')) {
      this.el.setAttribute('role', 'tablist');
    }
  }

  componentDidRender() {
    Promise.all([
      customElements.whenDefined('scale-tab-header'),
      customElements.whenDefined('scale-tab-panel'),
    ]).then(() => {
      this.linkPanels();
      this.propagateSizeToTabs();
    });

    if (this.small !== false) {
      statusNote({
        tag: 'deprecated',
        message:
          'Property "small" is deprecated. Please use the "size" property!',
        type: 'warn',
        source: this.el,
      });
    }
  }

  getAllTabs(): HTMLScaleTabHeaderElement[] {
    return Array.from(this.el.querySelectorAll('scale-tab-header'));
  }

  getAllEnabledTabs(): HTMLScaleTabHeaderElement[] {
    return Array.from(
      this.el.querySelectorAll('scale-tab-header:not([disabled])')
    );
  }

  getAllPanels() {
    return Array.from(this.el.querySelectorAll('scale-tab-panel'));
  }

  getPreviousTab() {
    const tabs = this.getAllEnabledTabs();
    const index = tabs.findIndex((tab) => tab.selected) - 1;
    return tabs[(index + tabs.length) % tabs.length];
  }

  getNextTab() {
    const tabs = this.getAllEnabledTabs();
    const index = tabs.findIndex((tab) => tab.selected) + 1;
    return tabs[index % tabs.length];
  }

  getFirstTab() {
    const tabs = this.getAllEnabledTabs();
    return tabs[0];
  }

  getLastTab() {
    const tabs = this.getAllEnabledTabs();
    return tabs[tabs.length - 1];
  }

  linkPanels() {
    const tabs = this.getAllEnabledTabs();
    const selectedTab = tabs.find((x) => x.selected) || tabs[0];

    tabs.forEach((tab) => {
      const panel = tab.nextElementSibling;
      tab.setAttribute('aria-controls', panel.id);
      panel.setAttribute('aria-labelledby', tab.id);
    });
    this.selectTab(selectedTab);
  }

  reset() {
    const tabs = this.getAllEnabledTabs();
    const panels = this.getAllPanels();

    tabs.forEach((tab) => (tab.selected = false));
    panels.forEach((panel) => (panel.hidden = true));
  }

  findPanelForTab(tab: HTMLScaleTabHeaderElement): HTMLScaleTabPanelElement {
    const panelId = tab.getAttribute('aria-controls');
    return this.el.querySelector(`#${panelId}`);
  }

  selectTab(nextTab: HTMLScaleTabHeaderElement) {
    const nextPanel = this.findPanelForTab(nextTab);

    this.reset();
    nextPanel.hidden = false;
    nextTab.selected = true;
  }

  /**
   * Sets or removes the `small` prop in `scale-tab-header` and `scale-tab-panel` children.
   */
  propagateSizeToTabs() {
    const action =
      this.size === 'small' || this.small ? 'setAttribute' : 'removeAttribute';
    const tabs = this.getAllTabs();
    const panels = this.getAllPanels();
    [...tabs, ...panels].forEach((child) => child[action]('small', ''));
  }

  render() {
    return (
      <Host>
        {this.styles && <style>{this.styles}</style>}

        <div part={this.getBasePartMap()} class={this.getCssClassMap()}>
          <slot name="tab" />
          <slot name="panel" />
        </div>
      </Host>
    );
  }

  getBasePartMap() {
    return this.getCssOrBasePartMap('basePart');
  }

  getCssClassMap() {
    return this.getCssOrBasePartMap('css');
  }

  getCssOrBasePartMap(mode: 'basePart' | 'css') {
    const component = 'tab-nav';
    const prefix = mode === 'basePart' ? '' : `${component}--`;

    return classNames(
      component,
      (this.size === 'small' || this.small) && `${prefix}small`
    );
  }
}
