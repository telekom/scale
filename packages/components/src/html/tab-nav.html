<!DOCTYPE html>
<html dir="ltr" lang="en">

<head>
	<meta charset="utf-8" />
	<meta name="viewport" content="width=device-width, initial-scale=1.0, minimum-scale=1.0, maximum-scale=5.0" />
	<title>Stencil Component Starter</title>

	<script type="module" src="/build/scale-components.esm.js"></script>
	<link rel="stylesheet" href="/build/scale-components.css" />
	<style type="text/css" media="screen, print">
		html {
			margin: 0;
			padding: 0;
		}

		body {
			margin: 0;
			padding: 4rem;
		}
	</style>
</head>

<body>
	<p>
		<scale-tab-nav>
			<scale-tab-header disabled slot="tab">General</scale-tab-header>
			<scale-tab-panel slot="panel">
				Freegan kinfolk farm-to-table humblebrag cred…
			</scale-tab-panel>
			<scale-tab-header slot="tab">Usage</scale-tab-header>
			<scale-tab-panel slot="panel">
				Bespoke austin pork belly yuccie pop-up. Before they sold out…
			</scale-tab-panel>
			<scale-tab-header selected slot="tab">Style</scale-tab-header>
			<scale-tab-panel slot="panel">
				Biodiesel chia af hoodie tumeric bespoke letterpress…
			</scale-tab-panel>
			<scale-tab-header slot="tab">Code</scale-tab-header>
			<scale-tab-panel slot="panel">
				Asymmetrical tattooed chia, banh mi blog microdosing…
			</scale-tab-panel>
		</scale-tab-nav>
	</p>
</body>

</html>