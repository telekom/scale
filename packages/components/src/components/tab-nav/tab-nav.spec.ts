/**
 * @license
 * Scale https://github.com/telekom/scale
 *
 * Copyright (c) 2021 Egor Kirpichev and contributors, Deutsche Telekom AG
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

import { newSpecPage } from '@stencil/core/testing';
import { TabNav } from './tab-nav';
import { TabHeader } from '../tab-header/tab-header';

describe('TabNav', () => {
  let page;
  beforeEach(async () => {
    page = await newSpecPage({
      components: [TabNav],
      html: `<scale-tab-nav></scale-tab-nav>`,
    });
  });

  it('should match snapshot', async () => {
    expect(page.root).toMatchSnapshot();
  });

  // This class seems silly but it's needed in order to make sure we query the element
  // since we can never be certain the name of the element will be <scale-tab-nav>.
  it('should have default scale-tab-nav class', async () => {
    expect(page.root.classList.contains('scale-tab-nav')).toBe(true);
  });

  it('should preselect tab having prop "selected" on tab-header', async () => {
    const pageWithTabs = await newSpecPage({
      components: [TabNav, TabHeader],
      html: `<scale-tab-nav>
              <scale-tab-header slot="tab">Tab 1</scale-tab-header>
              <scale-tab-panel slot="panel">
                Tab Content 1
              </scale-tab-panel>
              <scale-tab-header selected slot="tab">Tab 2</scale-tab-header>
              <scale-tab-panel slot="panel">
                Tab Content 2
              </scale-tab-panel>
            </scale-tab-nav>`,
    });
    const tabElements = pageWithTabs.root.querySelectorAll('scale-tab-header');
    expect(tabElements[0]).toEqualAttribute('aria-selected', 'false');
    expect(tabElements[1]).toEqualAttribute('aria-selected', 'true');
  });
});
