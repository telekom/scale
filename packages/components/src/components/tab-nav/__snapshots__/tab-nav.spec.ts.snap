// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabNav should match snapshot 1`] = `
<scale-tab-nav class="scale-tab-nav" role="tablist">
  <template shadowrootmode="open">
    <div class="tab-nav tab-nav--" part="tab-nav">
      <slot name="tab"></slot>
      <slot name="panel"></slot>
    </div>
  </template>
</scale-tab-nav>
`;
